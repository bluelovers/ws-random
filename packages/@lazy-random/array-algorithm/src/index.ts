
export * from './swapAlgorithm';
export * from './array_rebase';
