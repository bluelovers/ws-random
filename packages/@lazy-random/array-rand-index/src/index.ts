import { _MathRandom } from '@lazy-random/original-math-random';

export function arrayRandIndexByLength(len: number, ...argv: any[])
{
	return Math.floor(_MathRandom() * len)
}

export function arrayRandIndex(array: any[], ...argv: any[])
{
	return arrayRandIndexByLength(array.length)
}

export default arrayRandIndex
