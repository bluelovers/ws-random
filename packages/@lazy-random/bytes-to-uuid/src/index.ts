import { BYTE_TO_HEX_TO_LOWER_CASE } from '@lazy-random/shared-lib';

export function _createBytesToUuidFn(bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	return (buf: ArrayLike<number>, offset?: number) =>
	{
		let i = offset || 0;
		// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
		return ([
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
		]).join('');
	}
}

/**
 * @see https://github.com/kelektiv/node-uuid/blob/master/lib/bytesToUuid.js
 */
export function bytesToUuid(buf: ArrayLike<number>, offset?: number, bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	let i = offset || 0;
	// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
	return ([
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
	]).join('');
}

export default bytesToUuid
