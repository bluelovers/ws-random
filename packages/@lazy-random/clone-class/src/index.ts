/**
 * @todo support typescript
 */
export function getClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o;

	if (thisArgv instanceof RNGClass)
	{
		// @ts-ignore
		o = (thisArgv.__proto__.constructor)
	}
	else
	{
		o = RNGClass
	}

	return o
}

/**
 * @todo support typescript
 */
export function cloneClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o = getClass(RNGClass, thisArgv, ...argv)

	// @ts-ignore
	return new o(...argv)
}

export default cloneClass
