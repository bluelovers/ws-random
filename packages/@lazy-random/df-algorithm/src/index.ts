
export { dfBates } from './bates'

export { dfBernoulli } from './bernoulli'
export { dfBinomial } from './binomial'
export { dfExponential } from './exponential'
export { dfGeometric } from './geometric'

export { dfIrwinHall } from './irwin-hall'
export { dfLogNormal } from './log-normal'

export { dfNormal } from './normal'
export { dfPareto } from './pareto'
