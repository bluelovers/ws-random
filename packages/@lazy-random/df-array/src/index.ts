
export { dfArrayIndex } from './array-index'
export { dfArrayShuffle } from './array-shuffle'

export { dfArrayUnique } from './array-unique'
export { dfArrayFill } from './array-fill'

