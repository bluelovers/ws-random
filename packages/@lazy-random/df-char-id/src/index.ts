/**
 * Created by user on 2018/11/9/009.
 */

import UString from "uni-string";
import { ENUM_ALPHABET } from '@lazy-random/shared-lib';
import { expect } from '@lazy-random/expect';
import { floatToString } from '@lazy-num/float-to-string';
import { randIndex as _randIndex } from '@lazy-random/util-distributions';
import { IRNGLike } from '@lazy-random/rng-abstract';

export function dfCharID(random: IRNGLike, char?: ENUM_ALPHABET | string | Buffer | number, size?: number)
{
	if (typeof char === 'number')
	{
		if (typeof size === 'number')
		{
			char = floatToString(char)
		}
		else
		{
			[size, char] = [char, null];
		}
	}

	size = size || 8;
	//ow(size, ow.number.integer.gt(0));
	expect(size).integer.gt(0)

	if (!char)
	{
		char = ENUM_ALPHABET.DEFAULT
	}

	const ls = UString.create(char).split('');
	const len = ls.length;

	expect(ls).to.have.lengthOf.gt(1);

	const randIndex = () =>
	{
		return _randIndex(random, len)
	};

	return () =>
	{
		let i = size;
		let list: string[] = [];
		while (i--)
		{
			list.push(ls[randIndex()])
		}
		return list.join('');
	}
}

export default dfCharID
