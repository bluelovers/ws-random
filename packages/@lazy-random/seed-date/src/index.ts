import { _MathRandom } from '@lazy-random/original-math-random';
import { floatToString } from '@lazy-num/float-to-string';

type IFnRandomFloat = () => number;

export function seedFloatByDate(date: Date, fnRandomFloat: IFnRandomFloat)
{
	return date.valueOf() + (fnRandomFloat ?? _MathRandom)();
}

export function seedFloatByNow(fnRandomFloat?: IFnRandomFloat)
{
	return seedFloatByDate(new Date(), fnRandomFloat);
}

export function seedStringByDate(date: Date, fnRandomFloat?: IFnRandomFloat)
{
	return floatToString(seedFloatByDate(date, fnRandomFloat));
}

export function seedStringByNow(fnRandomFloat?: IFnRandomFloat)
{
	return floatToString(seedFloatByDate(new Date(), fnRandomFloat));
}
