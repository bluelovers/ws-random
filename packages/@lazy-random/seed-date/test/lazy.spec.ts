import { seedFloatByDate, seedStringByDate } from '../src/index';

describe(`test`, () =>
{
	const date = new Date('2020-01-01');
	const fnRandomFloat = () => 0.111111111111;

	([
		seedFloatByDate,
		seedStringByDate,
	]).forEach(fn =>
	{

		test(fn.name, () =>
		{

			let actual: any = fn(date, fnRandomFloat);

			expect(actual).toMatchSnapshot();

		});

	})

})
