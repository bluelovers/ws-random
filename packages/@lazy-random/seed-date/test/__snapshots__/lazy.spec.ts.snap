// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test seedFloatByDate 1`] = `1577836800000.111`;

exports[`test seedStringByDate 1`] = `"1577836800000.111083984375"`;
